driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -170,9 +188,15 @@ private void next(final SingleResultCallback<List<T>> callback, final boolean tr
         } else {
             ServerCursor localCursor = getServerCursor();
             if (localCursor == null) {
-                isClosed.set(true);
+                synchronized (this) {
+                    isClosed.set(true);
+                }
                 callback.onResult(null, null);
             } else {
+                synchronized (this) {
+                    isTrue("open", !isClosed.get() || closePending.get());
```

Comment:
`isTrue` is going to throw an exception, which will just get lost in an async context.  What is needed instead is something similar to line 178, followed by a `return` statement.

Reply:
Done